### ISSUE TITLE
Asynchronous updates of currentModelPath cause invalid state.save() results.

### ISSUE THREAD

**Issue by rerdavies** [mikeoliphant/neural-amp-modeler-lv2#36]
`currentModelPath `is updated asynchronously after the load has completed.  As a result, it is impossible for hosts to track both whether the plugin state has changed (and therefore whether the current patch needs to be saved), and what filename should be displayed in the UI. I believe that the correct behavior is to update `currentModelPath `before the worker thread is kicked off instead of after the load completes. 

The proposal is simple and straightforward: just update `currentModelPath `before starting the asynchronous load. A pull request to that effect will follow shortly.

The MOD convention: that plugins should send an unsolicited `PATCH_SET `message following a `state.restore()`. @falkTX says that the PATCH__Set message should occur only after a `state.restore()`, but in fact, the current implementation sends a patch_set message in both cases. (Changing that to `state.restore() `only doesn't actually help).

Asynchronous updating of state using the MOD convention makes it impossible for a host to support plugins that support the MOD convention, and those that do not;  and extremely difficult for hosts to decide whether plugin state has changed or not. PiPedal currently implements a "Pedalboard has changed" indicator, that lets the user know that pedal board settings have changed, and perhaps needs to be saved. I cannot for the life of me figure out how to do that with the current implementation, principally because of complications with asynchronous updating of `currentModelPath`. I'm faced with the prospect  of either:

- Delaying `state.save() `operations until `patch_Set` is received in order to accommodate the delayed update of state, and disabling patch_Get operations at startup because they cause aliased patch_Set messages. (The MOD convention)
 - OR tracking realtime `patch_Set` messages sent from the host in order to detect state changes, while using patch_Get to retrieve initial property values for UI components as they attach (which I think is the correct behavior)

I don't think it's actually possible to support both, if currentModelPath is updated asychronously; but entirely possible to support both if the proposed change is made.

As implemented, NAM currently sends a patch_Set message to the host after the load completes; but the host has no clean way to determine whether the state has changed, since it does not know whether the received patch_Set message was sent because of a State.restore(), or because of a patch_Set message sent to the plugin, or because of an intervening patch_Get message.

I have grave concerns about the MOD convention for sending patch_Set.  It is a non-standard extension to the LV2 standard. And it seems to me completely unnecessary. The host should be issuing patch_Get if and when UI attaches to the running plugin if it wants to know the default value of the property. The host can completely track state and UI updates without requiring plugins to follow the non-standard MOD convention, as long as they update their state synchronously, when receiving the patch_Set message.

My personal preference would be to provide support both for plugins that support the MOD convention; and plugins that don't.. But to do so, the host has to send patch_Get messages to retrieve the default value of the filename. Which of course sends patch_Set messages, which would be indistinguishable from the notification message that the MOD convention requires, further complicating state management of the asynchronous load update. This happens in the current provisional implementation of PiPedal code whenever NAM is select in the pedalboard, and a pedalboard is loaded from a saved pedalboard. The UI attaches and makes a PATCH_Get request before the asynchronous load has completed. So it's an actual concern rather than a theoretical concern.

Ironically, the current implementation puts host in precisely the situation that the patchChange notification is provided for: they have no way of knowing when state will update following either a patch.restore(), or a patch.load() except by tracking the extraneous patch_Set messages, and doing complex state tracking to decide after the fact whether the returned patch_Set message was caused by a state/restore(), or a sent patch._Set message.  The patch_Set mesage is, in fact, serving as a state__stateChanged message.

On the other hand, I think @falkTX  raised a valid point. patchChange is probably inappropriate, so long as state is updated synchronously. Pipedal has been modified to mark the state as dirty whenever a patch_Set from the host is executed on the audio thread. As a result, PiPedal no longer requires the plugin to send a stateChanged message. As long as the state is updated synchronously. For what it's worth, sending a stateChanged message would still make PiPedal do the right thing.

In passing, note that saving the state before the plugin is placed on the audio thread, and doing subsequent depthwise comparisons of current and saved state  does not provide a solution. As implemented, state.save() results will not reflect the results of a patch.restore() until the load completion notification is received on the audio thread.

**Comment by rerdavies** [comment]
A brief summary of the question at hand:

1) Should patch_Set values be reflected immediately in state, or should the change take effect asynchronously -- presumably only if the file loads properly. 

2) If asynchronously, how should the plugin notify the host that the change is now visible in state? I don't mind the spurious patch_Set message in the MOD implementation. But...

3) If using patch_Set to do that, how does a host distinguish between notification sent because of state.restore() which should not mark the state as unsaved,  patch_set (which would mark the state as unsaved), and patch_Gets sent by UI, which get stale values when the load is pending. 

4) If using patch_Set, should it be sent after state.load(), patch_Set or both? Almost definitely both, because otherwise hosts have no way to determine when calling state.save() is going to include the property change.

5) is it not better to require that the filename be reflected in state immediately when the load or state.restore() is started, which allows flexible implementations by hosts, and allows hosts to easily accommodate plugins that don't implement the MOD extension.

6) Is it possible to specify what hosts need to do to support plugins that update state asynchronously?  I think it's possible; but it's certainly complicated. And probably not possible to support this plugin as implemented and plugins that don't implement the MOD extension at the same time, because of the asynchronous state update issue.

The simplest solution, I think, is to update state immediately, in which case either approach works well, and the extra notification required by MOD is mostly harmless, whether it happens in response to the load or the load-complete.

It's kind of a peculiar issue. At the root of things, it requires PiPedal to ONLY support plugins that have implemented the MOD protocol, because of a peculiarity in the way that this plugin updates state. And wouldn't actually be an issue at all if this plugin also sent the stateChanged message. 

Seriously worth putting on the table: sending a STATE__stateChanged message would actually fix the problem. And one could actually argue that the core problem here is that PiPedal does NOT know when the state changes, because it's being updated asynchronously.  And is NOT able to distinguish between updates to state that were caused by a state.restore() and a patch_Set because the patch_Set reply as an indication of state being changes is aliased with other uses (patch_Get, particularly).

**Comment by falkTX** [comment]
you have 4 places where you posted issues or PRs, I dont even know where to start to give feedback... next time please consolidate and summarize things into a single place please, specially since it is all related.

you raise 2 points in #39 :
- write_current_path cannot be called during work_response.
- output atom sequence frame is not popped.

for the 1st, please show me where this is not allowed.
the worker response runs in the same "run" thread/context, in fact there is even a final worker end call that plugins can expose as needed for a final RT/run context handling.
so if you really think "popping" the frame is necessary (which is not true from my experience) we can do it doing the worker end call. that might even be the reason why that function exists in the first place...

from testing we know the current approach of not "popping" the frame works on all known hosts.
I have tried mod-host, carla and jalv. with others confirming it working in reaper and ardour.

also from https://github.com/lv2/lv2/blob/master/include/lv2/atom/forge.h#L180 we can see the "pop" call simply sets a local variable for the forge to know the writing has been finished. the actual port buffer is unchanged.
so while we can be pedantic and reintroduce this call, it is really not necessary at all.

on a similar tone, on #38 you write:

> the issue specifically is that `write_current_path` writes a message to the output forge, which is not in a valid state when handling work_response. 

and

> It is therefore invalid to use the atom_forge outside of process() call. The atom_forge will either be unintialized, or the sequence will be closed, depending on whether the scheduler callbacks occur before or after process().

this is untrue as seen running and working on all hosts so far.
the worker response happens during the run context, the buffers are still valid.

you state this is invalid as if it was an universal truth, but it is not what we observe happening.

one part is true, the sequence was closed before. but I intentionally do not close it now as to allow the worker response call to still write to it if it needs something extra during RT context.

now I can begin to reply to things on this ticket, will do it on a separate post

**Comment by falkTX** [comment]
> `currentModelPath `is updated asynchronously after the load has completed. As a result, it is impossible for hosts to track both whether the plugin state has changed (and therefore whether the current patch needs to be saved), and what filename should be displayed in the UI. I believe that the correct behavior is to update `currentModelPath `before the worker thread is kicked off instead of after the load completes.
> 
> The proposal is simple and straightforward: just update `currentModelPath `before starting the asynchronous load.

I disagree here.
If we change the "current model" before loading the actual model we will get inconsistent state.

A save action should save whatever the **current** plugin state is, not the pending one. (though for things that take a long time to load we could make an exception, but anyhow not the case here)

What happens if the model the user tried to load is invalid? Or it is valid but some operation (lack of memory, IO error, etc) made it fail to load.
It seems better to me to only store a new "current model" string after the model does indeed change. Otherwise it is inconsistent.

> The MOD convention: that plugins should send an unsolicited PATCH_SET message following a state.restore(). @falkTX says that the PATCH__Set message should occur only after a state.restore(), but in fact, the current implementation sends a patch_set message in both cases. (Changing that to state.restore() only doesn't actually help).

> Asynchronous updating of state using the MOD convention makes it impossible for a host to support plugins that support the MOD convention, and those that do not; and extremely difficult for hosts to decide whether plugin state has changed or not. PiPedal currently implements a "Pedalboard has changed" indicator, that lets the user know that pedal board settings have changed, and perhaps needs to be saved. I cannot for the life of me figure out how to do that with the current implementation, principally because of complications with asynchronous updating of currentModelPath

For keeping track of the state, I really do not see the issue.
Hosts that show a filename path/browser thing already have to keep track of which file is being set.
If they receive a patch-set message to the same path as the one already in place, there is nothing to do. If the path changes, they know a save is needed to keep the project state.
Is there a problem with this?

And really this is not "MOD convention" please stop with that idea.
The same approach used here is the same used in all DPF plugins, eg-sampler and OMK notes and many more.

Look at how eg-sampler, an official LV2 example plugin, does it.
In https://github.com/lv2/lv2/blob/master/plugins/eg-sampler.lv2/sampler.c#L232 it loads a file via worker and queues the file load notification via work response.
In https://github.com/lv2/lv2/blob/master/plugins/eg-sampler.lv2/sampler.c#L267 during work response it makes use of the forge to send a notification to the host that a new file was loaded.

this is exactly the same approach we are using here and in aidadsp-lv2 and many other projects.
note how eg-sampler never makes of "state changed" flag.

so it is not that this is a weird "MOD convention", it is that your implementation is non-standard.

to be honest I feel it is a bit pointless to continue here, as I see a lot of misconceptions happening in regards to what LV2 plugins should do.
also this can quickly get out of hand if trying to reply to everything...

so for now I would like to confirm that you understand these points then I can try to clarify the rest.

**Comment by rerdavies** [comment]
Thank you for the definitive sample. That helps immensely. In my self-defense, I would like to point out that LV2 documentation isn't particularly forthcoming on anything to do with this feature.

I think I have everything I need to fix pipedal, although I still have some concerns about obtaining initial state for the plugin that I THINK I can implement.

There is ONE substantial difference between the reference implementation, and yours. 

```
static LV2_State_Status
restore(LV2_Handle                  instance,
        LV2_State_Retrieve_Function retrieve,
        LV2_State_Handle            handle,
        uint32_t                    flags,
        const LV2_Feature* const*   features)
{
    ...
    if (!self->activated || !schedule) {
      // No scheduling available, load sample immediately
      lv2_log_trace(&self->logger, "Synchronous restore\n");
      Sample* sample = load_sample(&self->logger, path, self->sample_rate);
      if (sample) {
        free_sample(self, self->sample);
        self->sample         = sample;
        self->sample_changed = true;
      }
    } else {
        ... request scheduler.work ...
    }
    ...
}
```
And the subsequently: 
```
static void
run(LV2_Handle instance, uint32_t sample_count)
{
   ...
  // Send update to UI if sample has changed due to state restore
  if (self->sample_changed) {
    lv2_atom_forge_frame_time(&self->forge, 0);
    write_set_file(
      &self->forge, &self->uris, self->sample->path, self->sample->path_len);
    self->sample_changed = false;
  }
  ...
}
```

This has an important implication: namely, after calling state.restore(), state.save()) will return the same state data that was passed to state().restore(). This allows the host to make decisions about whether the pedalboard needs to be saved or not by comparing the current result of state.save() with the previous value.  

To be honest, that's the issue I've been struggling with: how do I tell whether state has changed without getting false positives? Has the state actually changed, or have I just run into the race condition between state.restore() and a PATCH_Set reply?

Note that the reference plugin does not have this problem; and yours does.

I think I can work around this; but I'd have to try it to be sure.

I am willing to accept that this pull request needs to be rejected. 

In the meantime, how do you feel about the following?

Either

1) implement the synchronous load if the plugin has not been activated. or

2) Just set the class currentFileName during patch..restore(), so that state.save() works, and let the load play itself out as before.

Option (2) is a small lie, but seems harmless to me.

Thoughts?

**Comment by rerdavies** [comment]
There are also a few other nagging details, like for instance whether every PATCH_SET must be followed by a PATCH_SET reply in order ensure that state gest saved, even if the property isn't a path? But I'm going to assume that they must for the sake of efficiency.

**Comment by rerdavies** [comment]
PiPedal has been updated to adhere to the conventions use by the LV2 sample plugin. 

Please reject pull request  #37, and close this issue. 

I'm not entirely sure where to document the host and plugin requirements. I don't these threads are indexed by search engines, and the will almost definitely not be indexed when the thread closes.  I'll give some thought as to where to post a guide. Maybe stackexchange.

**Comment by falkTX** [comment]
> Thank you for the definitive sample. That helps immensely. In my self-defense, I would like to point out that LV2 documentation isn't particularly forthcoming on anything to do with this feature.

Docs for LV2 have never been a strong point indeed. The example plugins tend to be good though, and cover most of the API.


> This has an important implication: namely, after calling state.restore(), state.save()) will return the same state data that was passed to state().restore(). This allows the host to make decisions about whether the pedalboard needs to be saved or not by comparing the current result of state.save() with the previous value.

there are 2 things here.

1. the lv2 example code keeps track of [de]activated state and uses it during restore to know if it should short-circuit the worker stuff and just load the state directly there. this doesn't apply to MOD because plugins are live/activated as soon as they are loaded, but still I think the "if deactivated load state now" approach makes sense.
so if that makes things easier on your side please do the changes to get that behaviour, I am not running PiPedal so don't have an easy way to test if it ends up working right or not

2. the `sample_changed` variable still feels a bit unnecessary for me. if you look closely at the lv2 example code that var is set to true even on the case of early/direct file load during restore. so the plugin ends up sending the loaded path to the host, that must be intentional and with the reasoning that the host/ui connected to the plugin needs to know the current loaded file path.
so while it matches your idea that we should have a var to know if a plugin loaded state or not, it is not for the same reason.
you want to avoid sending the file path back to the host after an initial restore, but the lv2 code uses the var to do exactly that (because there is a detection for deactivated state, bypassing the worker, meaning no final path reported to host)



> In the meantime, how do you feel about the following?
> 
> Either
> 
>     1. implement the synchronous load if the plugin has not been activated. or
> 
>     2. Just set the class currentFileName during patch..restore(), so that state.save() works, and let the load play itself out as before.
> 
> 
> Option (2) is a small lie, but seems harmless to me.

with the points above seems clear to me option 1 is the way to go.
but in a way that the plugin still reports the file path to the host at the end, like the eg-sampler does.

**Comment by falkTX** [comment]
> There are also a few other nagging details, like for instance whether every PATCH_SET must be followed by a PATCH_SET reply in order ensure that state gest saved, even if the property isn't a path? But I'm going to assume that they must for the sake of efficiency.

I dont follow what you mean here...
If a host sends a patch-set message, the plugin reads it and does not need to send back to the host again. (is it happening now? then it is unintentional)

from what I understand, what needs to happen is:
1. host tells plugin to load file, no reply needed or wanted (host assumes the file was loaded correctly)
2. plugin loads file without being directly told so (e.g. via state restore), then it must notify host of the path in use

do you agree with these 2 points?

**Comment by rerdavies** [comment]
@falxTX:

I agree on those two points. But I'm unclear on the question of whether PATCH_set ALWAYS requires a PATCH_set reply, _**even if the value is not a filename property.**_  

And I have a minor complaint about NAM reporting an invalid state between first load and the first PATCH_set reply (UNLIKE the reference plugin), which means that a host cannot reliably determine whether a pedalboard's state has changed since it was loaded.

For context, PiPedal has a UI indicator that indicates to the user that the current pedalboard preset has been changed, and might need saving. (A Save icon next to the preset selector, and a "*" after the name).

And things are somewhat complicate for me, in that attached web clients (there can be more than one), in order to maintain idempotence, post the entire pedalboard preset if plugins are added, removed, or connections are changed. This requires PiPedal to transmit state changes to connected web UIs when the state changes, instead of just doing a lazy update  on the server before saving. 

Here's what I've implemented.

- sending PATCH_Set to a plugin marks the current pedalboard as changed. This is occurs immediately, on the assumption that plugin state will be refreshed later, after a patch_Set reply is received. A message is sent from the realtime thread, and subsequently broadcast to all attached web clients to set the Unsaved Pedalboard flag.

- receiving a PATCH_Set reply on the realtime thread posts a message off the realtime thread, so that it can be broadcast to subscribers in web clients. That takes care of UI updates. An additional message Lv2StateMaybeChanged is sent if one or more PATCH_Sets have been received in a frame. This scheme minimizes the work that has to be done if the plugin has multiple fileName properties. TooB CabIR, for example, has three filename properties. This message avoids having to do three sequential state checks. 

- Offline, PATCH_set replies are cached, so that when a web client subscribes, it can be sent the last received value (if any). If the value has changed, the value is broadcast to subscribers, with a send/ack protocol that elides intermediate values if the web-clients isn't keeping up.

- If a MaybeLv2StateChange is received in the service thread, do a sate.save() and compare it to the previous results. If it has changed, broadcast the new state to all attached web clients. This requires taking a lock on the audio thread processing mutex.  Risky, but I see no alternative, since it can't be done without memory allocations (mapPath's use of `strdup/free` for example). Do NOT update Plugin Changed flag, since we may have picked up the invalid state that plugins like yours have between activation and first patch_Set.

The specific ambiguity I raised: what if the property isn't a filename? Is the plugin required to issue a PATCH_set to update state or not? No problem. PIpedal also gets and compares saved state when a PATCH_Set occurs, as well as when the PATCH_set reply is received. If the state doesn't change, no harm no foul. Even for NAM. In NAM's case, the indeterminate startup state will be saved and broadcast, but will shortly be overwritten by a valid state when patch_set is received. But can I really assume that every plugin in the world will issue a PATCH_set reply to indicate that state has changed, even if the property isn't a path? I think not.

This all works well for filename properties. But the PATCH_get/PATCH_set protocol is used for  more than just filename properties; it's also used by plugins to transmit state between the realtime thread and UI. 

Consider TooB Spectrograph, which makes PATCH_get/PATCH_set calls to get updates for spectral values. The Web UI issues PATCH_get, the plugin  schedules work to perform the FFT, and asynchronously  replies with PATCH_set containing 100 floating point values. 

In the case of TooB Spectrograph, everything works fine, because it doesn't have state. state.save() doesn't happen because the 
plugin doesn't support the state extension.  

But consider the case of a plugin that has both state AND patch properties that support UI.  In that case, each time the UI requests a state update, PiPedal will see a Patch_set reply, and has to make a risky state.save() call to make sure state hasn't changed. And the state.save() call has to take a mutex that has a significant chance of making the realtime thread underrun.  A risk that has to be taken 15 or 30 or 60 times a second. Or call state.save() without a mutex and run the risk of executing state.save() concurrently with plugin.run() -- which is, to tell the truth, the lesser of two evils that I chose.

And a similar problem with plugins that make calls that enable or disable features in the plugin when UI is attached. Toob Spetrograph, instead of using PATCH_Get to throttle spectrograph data, could have set a value to enable automatic sending of PATCH_Set replies containing spectrograph data. As implemented, this sets the pedalboard changed flag. [I suppose I could set a flag so that the pedalboard changed flag only gets set after a state change following a PATCH_set, further complicating an already fragile implementation. Probably will do that.]

And therein lies the main problem. There are collisions between use-cases for PATCH_Set Patch_Set reply that could have been trivially resolved if only the filename protocol requires use of state_stateChanged when the state changes asynchronously. But that ship has, apparently sailed.  In passing, use of state_stateChanged  would also have allowed reasonable use of a protective mutex around state.save().

In the meantime, the relevant specific complaint about NAM is that (UNLIKE the reference plugin) it has invalid state on first load between the time that state.restore() is called, and the time the first PATCH_set is received, which means that Pipedal can't compare state to decide whether state has changed.

**Comment by rerdavies** [comment]
> the sample_changed variable still feels a bit unnecessary for me.

If you look closely at the eg_sample code, you'll see that the synchronous load loads the sample directly; . Before activation, the output forge has not yet been initialized, and can't be written to. The flag is required so that the PATCH_set reply can be written. The PATCH_set must be written on initial load so that the UI knows what the value is. So the sample_changed variable is entirely neccessary. In the particular case of the sample code, the flag is used only for the unactivated patch_restore case. It could equally well be used in the work_complete() case as well, save for the fact that the PATCH_set reply would be delayed by one frame (so I won't do that).

```
 if (!self->activated || !schedule) {
    // No scheduling available, load sample immediately
    lv2_log_trace(&self->logger, "Synchronous restore\n");
    Sample* sample = load_sample(&self->logger, path, self->sample_rate);
    if (sample) {
      free_sample(self, self->sample);
      self->sample         = sample;
      self->sample_changed = true;
    }
  } else {
     ...
 }
```
A third option would be to set currentModelPath in patch_restore()/unactivated, and set a flag which will schedule a load of the initial model on the work thread on the first call to run(). Which honestly seems preferable because it means that the load will occur on the worker thread instead of a UI thread (or an asio websocket service thread in the case of PiPedal). And also doesn't require re-factoring of the current work() function.

Pull request forthcoming.

**Comment by falkTX** [comment]
> I _disagree_ on those two points. That takes care of updating UI; but that does not address the question of when the host must check to see whether state has changed. The plugin does current send PATCH_Set after PATCH_get when the load completes. The reference implemetnation does too. And this plugin must as well. Otherwise the host has no indication that the results of patch.save() may have changed (or what it has changed to).

I am not following on your reasoning for disagreeing. my 2 points were unrelated to patch-get requests, a host can be coded without the need for it in the first place. the patch-get method seems mostly useful to me as a way for an external UI to connect to an already loading plugin, as a means to query the current state.

When the host listens for patch-set messages (and caches the ones sent to the plugin), the host will always know the current state of the plugin assuming the plugin has no hidden state (which is not the case for NAM).

So I still do not see how this puts a barrier for knowing if the plugin state changed.

> I am also unclear on the question of whether PATCH_set ALWAYS requires a PATCH_set reply, even if the value is not a filename property. Must PiPedal do a full state.save() and compare every time a plugin sends UI state to it's UI? And can it really do so safely, given that state.save() is a non-realtime operation that must be executed while the realtime thread is running?

I think the reply is not really needed. Nothing in https://lv2plug.in/ns/ext/patch tells me that we need to send replies.
There might be some semantics for specific message types, nothing specific in https://lv2plug.in/ns/ext/atom#Path for that though.

What other values are we to consider for patch-set requests anyhow? I don't understand the confusion, the NAM plugin is ignoring all properties that are not of nam-model type, there is nothing else to do with other message types.

For whatever PiPedal needs to do for saving, it is up to you really. Saving a full state on each patch-set received on the host side seems too much, the host can simply cache the active value in use for each filename/path and compare against that one.

And finally yes, state save while plugin is running **must** be safe. It is only the restore that is potentially unsafe, and plugins can mark their state restore functionality as thread-safe with the use of https://lv2plug.in/ns/ext/state#threadSafeRestore

> And therein lies the main problem. There are collisions between use-cases for PATCH_Set Patch_Set reply that could have been trivially resolved if only the filename protocol requires use of state_stateChanged when the state changes asynchronously. But that ship has, apparently sailed. In passing, use of state_stateChanged would also have allowed reasonable use of a protective mutex around state.save().

> In the meantime, the relevant specific complaint about NAM is that (UNLIKE the reference plugin) it has invalid state on first load between the time that state.restore() is called, and the time the first PATCH_set is received, which means that Pipedal can't compare state to decide whether the pedalboard needs to be saved.

Let's work on these issues then.
The use of "stateChanged" is not the solution though, that flag is really meant for state that the host cannot keep track of. I hope it is clear by now that the host can and should keep track of filenames/paths, not just to show on its generic UI but also as a way to know if something changed. This is the typical approach of most hosts.

The use of mutex during save is just a compromise you imagined needed to be done - it is not needed and is unwanted.
If saving causes audio to temporarily stop, auto-save could never be a thing. If some plugin has issues doing save while processing audio at the same time something is wrong with that plugin and should be fixed.

With all that said I agree that we can make the plugin still better.
The case of loading the model file directly during restore when plugin is deactivated is a good one, though that makes the plugin not compatible with the "thread safe restore". For MOD it is preferable to have this "thread safe restore" way, but I can see in general it might not be desired (so if loading a project the first plugin run already has a valid model loaded)

I will comment on the PR you made next.

**Comment by rerdavies** [comment]
Let's just agree to disagree on whether stateChanged is neccessary and appropriate. I think it is. But the code has been rewritten to conform to your requirements at this point. I don't wish to discuss it further; and I would be grateful if you could just approve it, since it is exactly what you asked for.

> though that makes the plugin not compatible with the "thread safe restore"

The loading of the model file during restore does NOT prevent thread-safe restore, because the synchronous load is only performed if plugin.activate() has not been called.  If restore is performed after activation, the load is performed asynchronously, just as it was before. The pull request is completely compatible with "thread safe restore".

**Comment by falkTX** [comment]
it is not agree to disagree, state-changed was a late addition to lv2 when most plugins never needed such thing. it was only added in order to give plugins that have hidden (not host visible) state a chance to mark a project/session as dirty.

similar arguments were made in http://lists.lv2plug.in/pipermail/devel-lv2plug.in/2016-October/001691.html
the property that was added to lv2 was really just to let complex plugins tell the host that session is dirty and a save is needed to keep state as expected

if is not "my requirements", it is how LV2 basically works.
We have seen the official LV2 example code work on this fashion, I was hoping that was enough.
But ok moving on.

> The loading of the model file during restore does NOT prevent thread-safe restore, because the synchronous load is only performed if plugin.activate() has not been called.  If restore is performed after activation, the load is performed asynchronously, just as it was before. The pull request is completely compatible with "thread safe restore".

Indeed, I went to check because I was not 100% sure that activate() cannot be called while the state is being restored.

In https://lv2plug.in/c/html/group__state.html#a97532b93877a6bc3d7f88adcca6257a9 we can read about the restore function:

> This function is in the "Instantiation" threading class as defined by LV2. This means it MUST NOT be called concurrently with any other function on the same plugin instance. 

On https://lv2plug.in/ns/lv2core it puts [de]activate calls under "Instantiation Class" thread rules, so we are safe for this.

There are still some things on the PR that need a bit of tweaking, will comment on those a bit later

**Comment by rerdavies** [comment]
> Indeed, I went to check because I was not 100% sure that activate() cannot be called while the state is being restored.

You could just look at the sample plugin which uses precisely the same logic.

**Comment by rerdavies** [comment]
Resolved with commit https://github.com/mikeoliphant/neural-amp-modeler-lv2/commit/33a0c083278c2abed3a4ce202597d6ea4d5a8d7f 

Thank you.

**Comment by falkTX** [comment]
I can confirm from some quick testing it seems things are still working.
There is an issue loading LSTM models, often they produce invalid output, but that would be a different issue.

Thanks
